lib/logstash/outputs/logstash.rb

```diff
@@ -1,188 +1,281 @@
+
+    elsif @ssl_truststore_password
+      fail(LogStash::ConfigurationError, "`ssl_truststore_password` not allowed unless `ssl_truststore_path` is configured.")
+    end
   end
 
   def multi_receive(events)
     return if events.empty?
-    logger.trace("proxying #{events.size} events to inner HTTP plugin")
-    @internal_http.multi_receive(events)
-  rescue => e
-    logger.error("inner HTTP plugin has had an unrecoverable exception: #{e.message} at #{e.backtrace.first}")
-    raise
+
+    send_events(events)
   end
 
   def stop
-    logger.debug("stopping inner HTTP output plugin")
-    @internal_http.stop
-    logger.debug('inner HTTP output plugin has been stopped')
+    logger.debug("`logstash` output plugin has been stopped.")
   end
 
   def close
-    logger.debug("closing inner HTTP output plugin")
-    @internal_http.close
-    logger.debug('inner HTTP output plugin has been closed')
+    logger.debug("Closing `logstash` output plugin.")
+    http_client.close
+    logger.debug("`logstash` output plugin has been closed.")
   end
 
-  DEFAULT_PORT = 9800.freeze
+  private
 
-  def inner_http_output_options
-    @_inner_http_output_options ||= begin
-      http_options = {
-        'url'                  => construct_host_uri.to_s,
-        'http_method'          => 'post',
-        'retry_non_idempotent' => 'true',
-
-        # non-configurable codec
-        'content_type' => 'application/x-ndjson',
-        'format'       => 'json_batch',
-
-        'http_compression' => true,
-      }
-
-      if @username
-        http_options['user'] = @username
-        http_options['password'] = @password || fail(LogStash::ConfigurationError, '`password` is REQUIRED when `username` is provided')
-        logger.warn("transmitting credentials over non-secured connection") if @ssl_enabled == false
-      elsif @password
-        fail(LogStash::ConfigurationError, '`password` not allowed unless `username` is configured')
-      end
+  def construct_host_uri
+    scheme = @ssl_enabled ? 'https'.freeze : 'http'.freeze
+    @hosts.map do |destination| # Struct(:host,:port)
+      URI::Generic.build(:scheme => scheme,
+                         :host   => destination.host,
+                         :port   => destination.port || DEFAULT_PORT)
+    end.map(&:to_s).map(&:freeze)
+  end
 
-      if @ssl_enabled == false
-        rejected_ssl_settings = @original_params.keys.select { |k| k.start_with?('ssl_') } - %w(ssl_enabled)
-        fail(LogStash::ConfigurationError, "Explicit SSL-related settings not supported because `ssl_enabled => false`: #{rejected_ssl_settings}") if rejected_ssl_settings.any?
-      else
-        http_options['ssl_supported_protocols'] = @ssl_supported_protocols if @original_params.include?('ssl_supported_protocols')
+  def send_events(events)
+    body = LogStash::Json.dump(events.map(&:to_hash))
+    compressed_body = gzip(body)
+
+    loop do
```

Comment:
Our unit tests may become more clear if we break the contents of the loop out into a separate method so that we can validate the behavior of individual attempts and the retry/abort sequencing independently:

~~~ diff
diff --git a/lib/logstash/outputs/logstash.rb b/lib/logstash/outputs/logstash.rb
index 4068fa6..6cf98f3 100644
--- a/lib/logstash/outputs/logstash.rb
+++ b/lib/logstash/outputs/logstash.rb
@@ -165,25 +165,8 @@ class LogStash::Outputs::Logstash < LogStash::Outputs::Base
     compressed_body = gzip(body)
 
     loop do
-      url = ""
-      begin
-        response = @load_balancer.select do | selected_host_uri |
-          url = selected_host_uri
-          http_client.post(selected_host_uri, :body => compressed_body, :headers => @headers).call
-        end
-
-        break if response_success?(response.code)
-
-        retryable_response = retryable_response?(response.code)
-        log_response(url, response, events, retryable_response)
-
-        break unless retryable_response
-      rescue => exception
-        retryable_exception = retryable_exception?(exception)
-        log_exception(url, exception, body, retryable_exception)
-
-        break unless retryable_exception
-      end
+      next_action = transmit(body, compressed_body)
+      break unless next_action == :retry
 
       if pipeline_shutdown_requested?
         abort_batch_if_available!
@@ -199,6 +182,30 @@ class LogStash::Outputs::Logstash < LogStash::Outputs::Base
     raise e
   end
 
+  ##
+  # @param body [String]
+  # @param compressed_body [String]
+  # @return [:done, :abort, :retry]
+  def transmit(body, compressed_body)
+    url = nil
+    response = @load_balancer.select do | selected_host_uri |
+      url = selected_host_uri
+      http_client.post(selected_host_uri, :body => compressed_body, :headers => @headers).call
+    end
+
+    return :done if response_success?(response.code)
+
+    retryable_response = retryable_response?(response.code)
+    log_response(url, response, events, retryable_response)
+
+    return retryable_response ? :retry : :abort
+  rescue => exception
+    retryable_exception = retryable_exception?(exception)
+    log_exception(url, exception, body, retryable_exception)
+
+    return retryable_exception ? :retry : :abort
+  end
+
   def log_response(uri, response, events, retriable)
     response_code = response.code
     if retriable
~~~